server/src/main/java/org/elasticsearch/ingest/IngestSourceAndMetadata.java

```diff
@@ -0,0 +1,260 @@
+ */
+public class IngestSourceAndMetadata extends ValidatingMap {
+
+    private final ZonedDateTime timestamp;
+
+    private static final Map<String, BiFunction<String, Object, Object>> VALIDATORS = Map.of(
+        Metadata.INDEX.fieldName,
+        IngestSourceAndMetadata::stringValidator,
+        Metadata.ID.fieldName,
+        IngestSourceAndMetadata::stringValidator,
+        Metadata.ROUTING.fieldName,
+        IngestSourceAndMetadata::stringValidator,
+        Metadata.VERSION.fieldName,
+        IngestSourceAndMetadata::longValidator,
+        Metadata.VERSION_TYPE.fieldName,
+        IngestSourceAndMetadata::versionTypeValidator
+    );
+
+    public IngestSourceAndMetadata(Map<String, Object> map, ZonedDateTime timestamp) {
+        super(map, new HashMap<>(VALIDATORS));
+        this.timestamp = timestamp;
+    }
+
+    // These are available to scripts
+
+    public String getIndex() {
+        return getString(Metadata.INDEX.fieldName);
+    }
+
+    public void setIndex(String index) {
+        put(Metadata.INDEX.fieldName, index);
+    }
+
+    public String getId() {
+        return getString(Metadata.ID.fieldName);
+    }
+
+    public void setId(String id) {
+        put(Metadata.ID.fieldName, id);
+    }
+
+    public String getRouting() {
+        return getString(Metadata.ROUTING.fieldName);
+    }
+
+    public void setRouting(String routing) {
+        put(Metadata.ROUTING.fieldName, routing);
+    }
+
+    public String getVersionType() {
+        return getString(Metadata.VERSION_TYPE.fieldName);
+    }
+
+    public void setVersionType(String versionType) {
+        put(Metadata.VERSION_TYPE.fieldName, versionType);
+    }
+
+    public long getVersion() {
+        Number version = getNumber(Metadata.VERSION.fieldName);
+        assert version != null : Metadata.VERSION.fieldName + " validation allowed null version";
+        return version.longValue();
+    }
+
+    public void setVersion(long version) {
+        put(Metadata.VERSION.fieldName, version);
+    }
+
+    // timestamp isn't backed by the map
+    public ZonedDateTime getTimestamp() {
+        return timestamp;
+    }
+
+    // These are not available to scripts
+    public Number getIfSeqNo() {
+        return getNumber(Metadata.IF_SEQ_NO.fieldName);
+    }
+
+    public Number getIfPrimaryTerm() {
+        return getNumber(Metadata.IF_PRIMARY_TERM.fieldName);
+    }
+
+    @SuppressWarnings("unchecked")
+    public Map<String, String> getDynamicTemplates() {
+        return (Map<String, String>) mapValidator(Metadata.DYNAMIC_TEMPLATES.fieldName, get(Metadata.DYNAMIC_TEMPLATES));
+    }
+
+    /**
+     * Get Metadata value using the field name
+     */
+    public Object get(Metadata field) {
+        return get(field.fieldName);
+    }
+
+    /**
+     * One time operation that removes all metadata values and their validators and returns the underlying map, containing only source keys
+     */
+    public Map<String, Object> extractSource() {
```

Comment:
The existing pattern of use in `IngestDocument` is to remove the `Metadata` and then pass resulting map as the source.  The `IngestDocument` is [thrown away](https://github.com/elastic/elasticsearch/blob/d42211c4314aaeec0211f3cad40feea90c4ab1c5/server/src/main/java/org/elasticsearch/ingest/IngestService.java#L900..L920) so there's no need to allocate another, potentially large, map to hold the source to avoid modification.

Reply:
> Alternatively, what if source was kept separate, and delegated to.

We can do that, it would slightly complicate `EntrySet` but that's not a problem

Reply:
`source` and `metadata` maps are separated.